src/java.base/share/classes/com/sun/crypto/provider/DHKEM.java

```diff
@@ -0,0 +1,396 @@
+                return params.deriveKeyPair(ikm);
+            } catch (Exception e) {
+                throw new UnsupportedOperationException(e);
+            }
+        }
+
+        public KeyPair derive(int kem_id) {
+            Params params = Arrays.stream(Params.values())
+                    .filter(p -> p.kem_id == kem_id)
+                    .findFirst()
+                    .orElseThrow();
+            return derive(params);
+        }
+    }
+
+    private enum Params {
+
+        P256(0x10, 32, 32, 2 * 32 + 1,
+                "ECDH", "EC", CurveDB.lookup("secp256r1"), "SHA-256"),
+
+        P384(0x11, 48, 48, 2 * 48 + 1,
+                "ECDH", "EC", CurveDB.lookup("secp384r1"), "SHA-384"),
+
+        P521(0x12, 64, 66, 2 * 66 + 1,
+                "ECDH", "EC", CurveDB.lookup("secp521r1"), "SHA-512"),
+
+        X25519(0x20, 32, 32, 32,
+                "XDH", "XDH", NamedParameterSpec.X25519, "SHA-256"),
+
+        X448(0x21, 64, 56, 56,
+                "XDH", "XDH", NamedParameterSpec.X448, "SHA-512"),
+        ;
+
+        private final int kem_id;
+        private final int Nsecret;
+        private final int Nsk;
+        private final int Npk;
+        private final String kaAlgorithm;
+        private final String keyAlgorithm;
+        private final AlgorithmParameterSpec spec;
+        private final String hkdfAlgorithm;
+
+        private final byte[] suiteId;
+
+        Params(int kem_id, int Nsecret, int Nsk, int Npk,
+                String kaAlgorithm, String keyAlgorithm, AlgorithmParameterSpec spec,
+                String hkdfAlgorithm) {
+            this.kem_id = kem_id;
+            this.spec = spec;
+            this.Nsecret = Nsecret;
+            this.Nsk = Nsk;
+            this.Npk = Npk;
+            this.kaAlgorithm = kaAlgorithm;
+            this.keyAlgorithm = keyAlgorithm;
+            this.hkdfAlgorithm = hkdfAlgorithm;
+            suiteId = concat("KEM".getBytes(StandardCharsets.UTF_8),
+                    I2OSP(kem_id, 2));
+        }
+
+        private boolean isEC() {
+            return this == P256 || this == P384 || this == P521;
+        }
+
+        private KeyPair generateKeyPair(SecureRandom sr) {
+            if (sr instanceof RFC9180DeriveKeyPairSR r9) {
+                return r9.derive(this);
+            }
+            try {
+                KeyPairGenerator g = KeyPairGenerator.getInstance(keyAlgorithm);
+                g.initialize(spec, sr);
+                return g.generateKeyPair();
+            } catch (Exception e) {
+                throw new ProviderException(e);
+            }
+        }
+
+        private byte[] SerializePublicKey(PublicKey k) {
+            if (isEC()) {
+                ECPoint w = ((ECPublicKey) k).getW();
+                return ECUtil.encodePoint(w, ((NamedCurve) spec).getCurve());
+            } else {
+                byte[] uArray = ((XECPublicKey) k).getU().toByteArray();
+                return Arrays.copyOf(reverse(uArray), Npk);
```

Comment:
If I understand what your saying, if `Npk` is larger than `uArray`, then `copyOf()` will pad with zeros at the end.  I think that would change the value.  And I don't see why `Npk` would be smaller than `uArray` otherwise this would be returning an invalid key